li-apache-kafka-clients/src/main/java/com/linkedin/kafka/clients/consumer/LiKafkaFederatedConsumerImpl.java

```diff
@@ -137,7 +144,30 @@ public void subscribe(Collection<String> topics, ConsumerRebalanceListener callb
 
   @Override
   public void assign(Collection<TopicPartition> partitions) {
-    throw new UnsupportedOperationException("Not implemented yet");
+    if (partitions == null) {
+      throw new IllegalArgumentException("Topic partition collection to assign to cannot be null");
+    }
+
+    // If partitions are empty, it should be treated the same as unsubscribe().
+    if (partitions.isEmpty()) {
+      unsubscribe();
+      return;
+    }
+
+    Map<TopicPartition, ClusterDescriptor> topicPartitionToClusterMap =
+        _mdsClient.getClustersForTopicPartitions(_clientId, partitions, _mdsRequestTimeoutMs);
+
+    // Reverse the map so that we can have per-cluster topic partition sets.
+    Map<ClusterDescriptor, Set<TopicPartition>> clusterToTopicPartitionsMap = new HashMap<>();
```

Comment:
what about TopicPartitions for which no cluster was found?

Reply:
Argh.. I initially assumed that Mario would filter them out, but now that we decided to throw an error for nonexistent topics/partitions, we should check.. My bad.